### ISSUE TITLE
Reduce layer-creation boilerplate via high-level factories

### ISSUE THREAD

**Issue by LeoBuron** [es-ude/OnDeviceTraining#107]
## Motivation

Creating a Linear layer currently requires ~13 lines of boilerplate per layer (see `experiments/MnistExperiment.c::buildModel`):
- weight data buffer allocation
- weight dims
- `tensorInitWithDistribution` (weight)
- `gradInitFloat` (weight)
- `parameterInit` (weight)
- the same five lines for bias (with `ZEROS`)
- `linearLayerInit(w, b, q, q, q, q)`

`MnistExperiment::buildModel` ends up with ~26 lines for only 2 Linear layers. The repetition is nearly mechanical — the only layer-specific information is `inFeatures`, `outFeatures`, and the distribution choice.

After PR #106 (initTensor + initDistribution + tensorFillFromBuffer), the building blocks for a one-liner factory exist and are clean; this issue is the natural follow-up.

## Goal

Instantiating a layer becomes a single line:

```c
model[0] = linearLayerInitFloat(784, 20);                    // all FLOAT32
model[2] = linearLayerInit(20, 10, fwQ, wgQ, bgQ, plQ);      // mixed-precision
```

Both variants allocate weights, bias, gradients, and parameters internally. The caller no longer touches buffers.

## Decisions from earlier scoping (2026-04-15)

- **Internal allocation is OK for userApi factories.** The "caller-owned memory" preference applies *outside* userApi. For layer factories specifically, `reserveMemory` inside the factory is acceptable.
- **Flexibility for all 4 quantizations remains required.** The verbose variant must let callers pass 4 separate `quantization_t *`.
- **Per-layer API may differ.** Work layer-by-layer — Linear first (biggest pain point).
- **Distribution defaults orient toward PyTorch where possible.** Bit-identical PyTorch defaults are blocked on a separate Distribution-parameters issue. Until then, document the actual default and the divergence.
- **No Sequential Builder.** One-liner factories + a manual `layer_t *model[]` array are sufficient.
- **Boilerplate should shrink in the verbose variant too**, not only in the Float convenience wrapper.

## Design (tentative — refine during implementation)

For each existing layer type (Linear, ReLU, Softmax), add a `*LayerInitFloat(...)` one-liner factory that internally:

- allocates weight/bias buffers via `reserveMemory`
- initializes weights with the layer's default distribution
- initializes bias with `ZEROS`
- builds `parameter_t` + grad tensors
- constructs and returns `layer_t *`

The verbose API is reconsidered during implementation. Two live options:

- **Option A (additive):** keep current 6-parameter `linearLayerInit(parameter_t *w, parameter_t *b, 4× quant)` for users who build parameters themselves.
- **Option B (replace):** change to `linearLayerInit(inF, outF, 4× quant)` — same internal allocation as the Float variant, per-quant-slot control. Breaking change; boilerplate collapses to one line even for mixed precision.

Option B better matches the "less boilerplate even in the verbose variant" intent. The trade-off is loss of "user builds their own `parameter_t`" use cases — audit shows `MnistExperiment.c` is the only current caller. Likely safe.

## Parameter ownership (deferred from PR #106)

PR #106 closed the H3 segfault in `freeParameter` via a NULL-grad guard, but left the asymmetric ownership of non-trainable parameters open: `linearLayerInitNonTrainable` constructs `parameter_t` instances *internally*, but `freeLinearLayer` only frees the config struct chain — the parameters are leaked. Trainable parameters, by contrast, come from the caller and are freed by the optimizer (`SgdApi.c:87`).

This factory rewrite is the natural moment to make that ownership story first-class: layers built by `*LayerInitFloat` should own and free their parameters; layers built by an explicit-parameter API (Option A or current code) should not. The chosen verbose-variant shape (A vs B) directly determines the answer.

## Scope

**In scope:**
- `linearLayerInitFloat`, `reluLayerInitFloat`, `softmaxLayerInitFloat`
- Matching verbose (mixed-precision) signatures for each
- Refactor `MnistExperiment.c::buildModel` to use the new API
- Unit tests for each factory (layer_t correctness, parameter dims, grad allocation, forward pass smoke)
- Decide and implement the parameter-ownership model — close the leak from `linearLayerInitNonTrainable`

**Out of scope:**
- PyTorch-bit-identical distribution defaults (separate issue)
- Flatten one-liner (separate issue)
- Mixed-precision convenience variants like `*LayerInitSymInt32`, `*LayerInitAsym`

## Open questions to resolve during implementation

- Exact verbose signature: `(inF, outF, 4× quant)` or `(parameter_t *, parameter_t *, 4× quant)`?
- If both: what are the names? (`linearLayerInit` vs `linearLayerInitWithParams` vs something else)
- Should `reluLayerInitFloat()` take zero args or one `quantization_t *`?
- Which default distribution per layer? (Xavier-Uniform gain=1 until distribution-params issue lands; maybe Kaiming for layers before ReLU.)
- Trainable/non-trainable distinction: implicit via "factory variant chosen", or explicit field on `layer_t`?

## Test plan

- Factory returns `layer_t` of the correct type
- Weight/bias shapes match `inF × outF` / `1 × outF`
- Gradients allocated with matching shapes
- Forward pass on a hand-constructed input produces the expected output shape
- Mutation test: breaking the distribution init or grad init inside the factory causes a test failure
- `freeLinearLayer` (or its successor) on a `*LayerInitFloat`-built layer leaves zero leaked allocations under valgrind
- End-to-end: `MnistExperiment` runs with the new API and converges to comparable accuracy
